torch/_inductor/freezing.py

```diff
@@ -0,0 +1,208 @@
+import itertools
+import weakref
+from typing import List, Optional, Tuple
+
+import torch
+import torch.utils._pytree as pytree
+from . import config
+
+
+def replace_node_with_constant(gm, node, constant):
+    g = gm.graph
+
+    i = 0
+    while True:
+        qualname = f"_frozen_param{i}"
+        if not hasattr(gm, qualname):
+            break
+        i += 1
+
+    with g.inserting_before(node):
+        new_input_node = g.create_node("get_attr", qualname, (), {})
+        node.replace_all_uses_with(new_input_node)
+        new_input_node.meta.update(node.meta)
+        g.erase_node(node)
+
+    # needed to suppress `does not reference an nn.Module, nn.Parameter, or buffer` warning
+    gm.register_buffer(qualname, constant)
+    setattr(gm, qualname, constant)
+
+
+def replace_params_with_constants(gm, flat_params, fw_metadata) -> List[int]:
+    """
+    Replaces the parameters of a PyTorch GraphModule with constants wherever possible.
+
+    Returns a list of indices representing the input parameters that were not converted to constants.
+    """
+
+    params = [node for node in gm.graph.nodes if node.op == "placeholder"]
+    fake_inp_nodes = [node for (_, node) in zip(flat_params, params)]
+
+    g = gm.graph
+
+    preserved_arg_indices = []
+    aliased_input_args = [
+        out_info.base_idx
+        for out_info in fw_metadata.output_info
+        if out_info.base_idx is not None
+    ]
+
+    for i, (real_input, node) in enumerate(zip(flat_params, fake_inp_nodes)):
+        if i in fw_metadata.mutated_inp_indices or aliased_input_args:
+            preserved_arg_indices.append(i)
+            continue
+
+        replace_node_with_constant(gm, node, real_input)
+
+    # add on non param inputs
+    preserved_arg_indices.extend(range(len(flat_params), len(params)))
+
+    # is this necessary ?
+    gm.recompile()
+    return preserved_arg_indices
+
+
+@torch.utils._python_dispatch._disable_current_modes()
+def constant_fold(gm):
+    unknown_value = object()
+
+    node_replacements = {}
+
+    class ConstantFolder(torch.fx.Interpreter):
```

Comment:
Will this work for dynamic shapes?

I am thinking something like `torch.zeros([SymInt, SymInt])` might get constant folded?

Also, thoughts on if we should constant fold `zeros`, `ones`, `full`, etc?
1) If they are used by other constant ops -- especially fallback ones or params, this seems like a good folding to do
2) If they are used by non-constant ops -- inductor will just turn them into a constant in the generated kernel so folding will result in an extra load.

Reply:
Good points. I'll double check that Symbolic shapes works. 

For `zeros`, `ones`, `full` etc - we could check uses, we could also add logic so that constants with a uniform value get mapped to `Constant`. I think the latter would be simpler. We can pass along that certain constants are uniform in the meta.

Reply:
There is also `torch.arange` which inductor can codegen without any load.

I think we likely don't want to constant fold these, because I suspect inductor generates near optimal code for them.   Though there certainly are exceptions.  Might be worth benchmarking.

Reply:
I just skipped constructors for now, will come back to this and benchmark different approaches